solver/container.go

```diff
@@ -121,6 +183,62 @@ func getExitCode(err error) (uint8, error) {
 	return 0, err
 }
 
+func (s *Solver) forwardProgress(id digest.Digest, lg *zerolog.Logger) (progress.Writer, io.WriteCloser, io.WriteCloser, func()) {
+	reader, readerCtx, stopReader := progress.NewContext(context.Background())
+	// for some reason, stopReader is a cancel associated with a different context,
+	// so we have to use both that and a separate readerCancel created here
+	readerCtx, readerCancel := context.WithCancel(readerCtx)
+
+	writer, _, writerCtx := progress.NewFromContext(readerCtx, progress.WithMetadata("vertex", id))
+
+	stdout, stderr, flushLogs := logs.NewLogStreams(writerCtx, false)
+
+	cancelAll := func() {
+		writer.Close()
+		flushLogs()
+		stopReader()
+		readerCancel()
+	}
+
+	s.eventsWg.Add(1)
+	go func() {
+		defer s.eventsWg.Done()
+
+		for {
+			pgresses, err := reader.Read(readerCtx)
+			if err != nil {
+				if !errors.Is(err, context.Canceled) && !errors.Is(err, io.EOF) {
+					lg.Error().Err(err).Msg("failed to read progress")
+				}
+				return
+			}
+			statuses := &bk.SolveStatus{}
+			for _, pgress := range pgresses {
+				switch v := pgress.Sys.(type) {
+				case bk.Vertex:
+					statuses.Vertexes = append(statuses.Vertexes, &v)
+				case bk.VertexLog:
+					vtx, ok := pgress.Meta("vertex")
+					if !ok {
+						lg.Debug().Msg("failed to find vertex in progress")
+						continue
+					}
+					v.Vertex = vtx.(digest.Digest)
+					v.Timestamp = pgress.Timestamp
+					statuses.Logs = append(statuses.Logs, &v)
+				}
+			}
+			select {
+			case <-readerCtx.Done():
+				return
+			case s.opts.Events <- statuses:
```

Comment:
> Since they are also a source of events in solver.SolveRequest, are we sure that we will not hit the issue where events are not fully sent?
> Is it accurate to also use the waiting group here to be sure?

I think this all works out, but there are a lot of various components involved. To break it down:
1. Here we do add before the goroutine starts, done only when the goroutine exits
2. That means that `solver.Stop` will block until this goroutine has exited ([see here](https://github.com/sipsma/dagger/blob/68769bad36932436f2fade0427cbba771d0e7eb7/solver/solver.go#L90-L90)). This goroutine will only exit when the container itself has been stopped (whether through `core.#Stop` or because dagger is shutting down). In either case that means we are done reading output from it.
3. `solver.Stop` is called [here](https://github.com/sipsma/dagger/blob/68769bad36932436f2fade0427cbba771d0e7eb7/client/client.go#L238-L238) in the `buildfn`, which in turn is called as part of this separate waitgroup in `client.Do` [here](https://github.com/sipsma/dagger/blob/68769bad36932436f2fade0427cbba771d0e7eb7/client/client.go#L124-L124).
4. That separate waitgroup ensures that even if `buildfn` exits, we will still wait for logs to be flushed to the output before exiting.

So I don't think we can hit a situation where events aren't fully sent. The only thing I can think of is that we might be able to see logs saying that the container was stopped before all the stdout/stderr logs are flushed, so the ordering would look odd. I'm not sure this is a huge deal, but I'll think about whether there is an easy way to fix it.

> Another questions: logs will be sent in parallel, will it reduce the visibility of other logs?

Yes they will but unless there's some mechanism I'm unaware of, this should be the case with `core.#Exec`s running in parallel with one another too.

Reply:
I tested more and couldn't see that the stop completed message ever showed up before the logs were flushed and written, so I think no issue there either.